jerry-core/ecma/builtin-objects/ecma-builtin-symbol.c

```diff
@@ -0,0 +1,239 @@
+ #include "ecma-literal-storage.h"
+ #include "ecma-try-catch-macro.h"
+ #include "jcontext.h"
+ #include "jrt.h"
+
+#ifndef CONFIG_DISABLE_ES2015_SYMBOL_BUILTIN
+
+#define ECMA_BUILTINS_INTERNAL
+#include "ecma-builtins-internal.h"
+
+#define BUILTIN_INC_HEADER_NAME "ecma-builtin-symbol.inc.h"
+#define BUILTIN_UNDERSCORED_ID symbol
+#include "ecma-builtin-internal-routines-template.inc.h"
+
+/** \addtogroup ecma ECMA
+ * @{
+ *
+ * \addtogroup ecmabuiltins
+ * @{
+ *
+ * \addtogroup symbol ECMA Symbol object built-in
+ * @{
+ */
+
+/**
+ * Handle calling [[Call]] of built-in Symbol object.
+ *
+ * @return ecma value
+ */
+ecma_value_t
+ecma_builtin_symbol_dispatch_call (const ecma_value_t *arguments_list_p, /**< arguments list */
+                                   ecma_length_t arguments_list_len) /**< number of arguments */
+{
+  JERRY_ASSERT (arguments_list_len == 0 || arguments_list_p != NULL);
+
+  return ecma_op_create_symbol (arguments_list_p, arguments_list_len);
+} /* ecma_builtin_symbol_dispatch_call */
+
+/**
+ * Handle calling [[Construct]] of built-in Symbol object.
+ *
+ * Symbol constructor is not intended to be used
+ * with the new operator or to be subclassed.
+ *
+ * See also:
+ *          ECMA-262 v6, 19.4.1
+ * @return ecma value
+ */
+ecma_value_t
+ecma_builtin_symbol_dispatch_construct (const ecma_value_t *arguments_list_p, /**< arguments list */
+                                        ecma_length_t arguments_list_len) /**< number of arguments */
+{
+  JERRY_ASSERT (arguments_list_len == 0 || arguments_list_p != NULL);
+
+  return ecma_raise_type_error (ECMA_ERR_MSG ("Symbol is not a constructor."));
+} /* ecma_builtin_symbol_dispatch_construct */
+
+/**
+ * Helper function for Symbol object's 'for' and `keyFor`
+ * routines common parts
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static ecma_value_t
+ecma_builtin_symbol_for_helper (ecma_value_t value_to_find) /**< symbol or ecma-string */
+{
+  ecma_string_t *string_p;
+
+  bool is_for = ecma_is_value_string (value_to_find);
+
+  if (is_for)
+  {
+    string_p = ecma_get_string_from_value (value_to_find);
+  }
+  else
+  {
+    string_p = ecma_get_symbol_from_value (value_to_find);
+  }
+
+  ecma_lit_storage_item_t *symbol_list_p = JERRY_CONTEXT (symbol_list_first_p);
+  jmem_cpointer_t *empty_cpointer_p = NULL;
+
+  while (symbol_list_p != NULL)
+  {
+    for (int i = 0; i < ECMA_LIT_STORAGE_VALUE_COUNT; i++)
+    {
+      if (symbol_list_p->values[i] != JMEM_CP_NULL)
+      {
+        ecma_string_t *value_p = JMEM_CP_GET_NON_NULL_POINTER (ecma_string_t,
+                                                               symbol_list_p->values[i]);
+
+        if (is_for)
+        {
+          ecma_string_t *symbol_desc_p = ecma_get_symbol_description (value_p);
+
+          if (ecma_compare_ecma_strings (symbol_desc_p, string_p))
+          {
+            ecma_deref_ecma_string (symbol_desc_p);
```

Comment:
The reason is [this](https://github.com/rerobika/jerryscript/blob/symbols/jerry-core/ecma/builtin-objects/ecma-builtin-symbol.c#L194) line. Since `ecma_op_to_string` increases the strings ref count or allocates a new string for non-string values. Also this `free` step could be done in the end of `ecma_builtin_symbol_for` but during the comparison it's easier to do it, since the string is not wrapped to an `ecma_value_t`.